# Assignment: Data Wrangling

In this assignment, we will practice data wrangling techniques on real world data.

## NYC film permits

New York City is a very popular film/tv shooting location. In order to assert the exclusive usage of city properties (e.g. sidewalk, park, etc.) for this purpose, a permit is required. New York Mayor's Office of Media and Entertainment release all granted [film permits data](https://data.cityofnewyork.us/City-Government/Film-Permits/tg4x-b46p) to the public. In this part, we will do some data wrangling on this data.

Let us first retrieve the data!

In [13]:
import pandas as pd

url = "https://data.cityofnewyork.us/api/views/tg4x-b46p/rows.csv"
    
film_permits = pd.read_csv(url)

Your task for this part is to clean up this data. Specifically, the cleaned up dataframe should have the following properties:
1. All columns must have the correct data type.  In particular, `StartDateTime`, `EndDateTime`, `EnteredOn` should have `datetime64` as data type.
2. The `ZipCode(s)` column may contain multiple zip codes in one cell. In the cleaned up dataframe, we will replace the `ZipCode(s)` column with `ZipCode` column, where each cell in the `ZipCode` column only contains a single zip code. Rows with multiple zip codes in the input data frame should be repeated for each zip code in the output dataframe.
3. The output dataframe should not contain any missing data.

In [28]:
# Your code here...

# 1. Changing the data type
film_permits['StartDateTime'] = pd.to_datetime(film_permits['StartDateTime'])
film_permits['EndDateTime'] = pd.to_datetime(film_permits['EndDateTime'])
film_permits['EnteredOn'] = pd.to_datetime(film_permits['EnteredOn'])

# 2. Handling the zipcode(s) column to replace it with a single ZipCode column
# Rows with multiple zip codes in the input data frame is repeated for each zip code in the output dataframe
film_permits = film_permits.assign(ZipCode=film_permits['ZipCode(s)'].str.split(',')).explode('ZipCode')

# 3. Removing rows with missing data
cleaned_film_permits = film_permits.dropna()

In [29]:
cleaned_film_permits

EventID                      EventType       StartDateTime  \
0      696255                Shooting Permit 2023-02-17 09:00:00   
1      714139                Shooting Permit 2023-05-12 13:00:00   
2      705334                Shooting Permit 2023-04-10 09:00:00   
2      705334                Shooting Permit 2023-04-10 09:00:00   
2      705334                Shooting Permit 2023-04-10 09:00:00   
...       ...                            ...                 ...   
7866   793926                Shooting Permit 2024-06-20 09:00:00   
7866   793926                Shooting Permit 2024-06-20 09:00:00   
7867   793783  Theater Load in and Load Outs 2024-06-20 00:01:00   
7868   785081  Theater Load in and Load Outs 2024-06-20 00:01:00   
7869   792643                Shooting Permit 2024-06-20 07:00:00   

             EndDateTime           EnteredOn  \
0    2023-02-18 12:00:00 2023-02-14 22:47:33   
1    2023-05-13 05:00:00 2023-05-04 14:27:51   
2    2023-04-10 22:00:00 2023-03-30 17:17:29   
2    2023-04-10 22:00:00 2023-03-30 17:17:29   
2    2023-04-10 22:00:00 2023-03-30 17:17:29   
...                  ...                 ...   
7866 2024-06-20 22:00:00 2024-06-17 22:24:41   
7866 2024-06-20 22:00:00 2024-06-17 22:24:41   
7867 2024-06-24 23:59:00 2024-06-17 13:21:55   
7868 2024-06-21 06:00:00 2024-05-16 12:54:50   
7869 2024-06-20 22:00:00 2024-06-12 17:10:02   

                                  EventAgency  \
0     Mayor's Office of Media & Entertainment   
1     Mayor's Office of Media & Entertainment   
2     Mayor's Office of Media & Entertainment   
2     Mayor's Office of Media & Entertainment   
2     Mayor's Office of Media & Entertainment   
...                                       ...   
7866  Mayor's Office of Media & Entertainment   
7866  Mayor's Office of Media & Entertainment   
7867  Mayor's Office of Media & Entertainment   
7868  Mayor's Office of Media & Entertainment   
7869  Mayor's Office of Media & Entertainment   

                                            ParkingHeld    Borough  \
0     KINGSLAND AVENUE between DEAD END and GREENPOI...   Brooklyn   
1     WEST   26 STREET between 12 AVENUE and 11 AVEN...  Manhattan   
2     SOUTH STREET between BROAD STREET and OLD SLIP...  Manhattan   
2     SOUTH STREET between BROAD STREET and OLD SLIP...  Manhattan   
2     SOUTH STREET between BROAD STREET and OLD SLIP...  Manhattan   
...                                                 ...        ...   
7866  SCHERMERHORN STREET between COURT STREET and B...   Brooklyn   
7866  SCHERMERHORN STREET between COURT STREET and B...   Brooklyn   
7867  WEST   35 STREET between 8 AVENUE and 9 AVENUE...  Manhattan   
7868  FROST STREET between DEBEVOISE AVENUE and MORG...   Brooklyn   
7869  COLUMBUS AVENUE between WEST   62 STREET and W...  Manhattan   

     CommunityBoard(s) PolicePrecinct(s)           Category  SubCategoryName  \
0                    1                94         Television   Cable-episodic   
1                    4                10         Television  Episodic series   
2                 1, 3              1, 5               Film          Feature   
2                 1, 3              1, 5               Film          Feature   
2                 1, 3              1, 5               Film          Feature   
...                ...               ...                ...              ...   
7866           2, 3, 4    79, 81, 83, 84  Still Photography   Not Applicable   
7866           2, 3, 4    79, 81, 83, 84  Still Photography   Not Applicable   
7867                 4                14            Theater          Theater   
7868                 1                94            Theater          Theater   
7869              4, 7            18, 20         Television  Episodic series   

                       Country                  ZipCode(s) ZipCode  
0     United States of America                       11222   11222  
1     United States of America                       10001   10001  
2     United S

Once the data frame is clean, please answer the following questions with code:
1. Which borough has the most _unique_ events for television, film and theater respectively?
2. Which zip code has the most _unique_ events for television, film and theater respectively?

In [42]:
# Your code here...

# Group by 'Borough' and 'Category' to find the borough with the most unique events for TV, film, and theater
borough_event_counts = cleaned_film_permits.groupby(['Borough', 'Category']).size().reset_index(name='EventCount')
borough_most_unique_tv = borough_event_counts[borough_event_counts['Category'].str.contains('Television', case=False)].sort_values('EventCount', ascending=False).head(1)
borough_most_unique_film = borough_event_counts[borough_event_counts['Category'].str.contains('Film', case=False)].sort_values('EventCount', ascending=False).head(1)
borough_most_unique_theater = borough_event_counts[borough_event_counts['Category'].str.contains('Theater', case=False)].sort_values('EventCount', ascending=False).head(1)

# Group by 'ZipCode' and 'Category' to find the zip code with the most unique events for TV, film, and theater
zipcode_event_counts = cleaned_film_permits.groupby(['ZipCode', 'Category']).size().reset_index(name='EventCount')
zipcode_most_unique_tv = zipcode_event_counts[zipcode_event_counts['Category'].str.contains('Television', case=False)].sort_values('EventCount', ascending=False).head(1)
zipcode_most_unique_film = zipcode_event_counts[zipcode_event_counts['Category'].str.contains('Film', case=False)].sort_values('EventCount', ascending=False).head(1)
zipcode_most_unique_theater = zipcode_event_counts[zipcode_event_counts['Category'].str.contains('Theater', case=False)].sort_values('EventCount', ascending=False).head(1)

# Display results
print("\nBorough with the most unique TV events:\n", borough_most_unique_tv)
print("\nBorough with the most unique Film events:\n", borough_most_unique_film)
print("\nBorough with the most unique Theater events:\n", borough_most_unique_theater)

print("\nZip code with the most unique TV events:\n", zipcode_most_unique_tv)
print("\nZip code with the most unique Film events:\n", zipcode_most_unique_film)
print("\nZip code with the most unique Theater events:\n", zipcode_most_unique_theater)


Borough with the most unique TV events:
       Borough    Category  EventCount
23  Manhattan  Television        2535

Borough with the most unique Film events:
       Borough Category  EventCount
19  Manhattan     Film         827

Borough with the most unique Theater events:
       Borough Category  EventCount
24  Manhattan  Theater        1780

Zip code with the most unique TV events:
     ZipCode    Category  EventCount
970   11222  Television         778

Zip code with the most unique Film events:
     ZipCode Category  EventCount
575   10002     Film          71

Zip code with the most unique Theater events:
     ZipCode Category  EventCount
679   10019  Theater         295


## Answer

### Student name:
MANALI TANNA

### Your answer:
1. The borough with the most unique TV events is Manhattan
2. The borough with the most unique film events is Manhattan
3. The borough with the most unique theater events is Manhattan

4. The zip code with the most unique TV events is 11222
5. The zip code with the most unique film events is 10002
6. The zip code with the most unique theater events is 10019